### DISCUSSION TITLE
embedded sketchpad

### DISCUSSION THREAD

**Discussion by baptiste** [quarto-dev/quarto-cli#4391] [Feature Requests]
I think it'd be very useful to be able to make hand-drawn figures right within the editor, by embedding something like [tl;draw](tldraw.com/) or [excalidraw](excalidraw.com/). Both can store the sketch strokes as json/svg for further editing, and export either png/svg output. 

![6e3b2c1519e4e91b](https://user-images.githubusercontent.com/18970/219457925-a120e9fd-0efc-4be3-92c4-a45352f384e7.png)

(this is just a mockup)

[draw.io/](draw.io/) comes close to achieving what I mean, in VScode at least, as you can have one panel open to edit the drawio file, while another panel is showing directly the output embedded in a qmd document. 

![Screenshot 2023-02-17 at 08 25 51](https://user-images.githubusercontent.com/18970/219467096-9bcf0dae-53ad-4878-b34a-62a9e70233c5.png)

With the {minidrawio} and {devoutdrawio} packages by @coolbutuseless, one can easily generate a graphic first, then edit it at will interactively and have the graphic embedded in quarto update in real-time.

![Screenshot 2023-02-18 at 17 48 07](https://user-images.githubusercontent.com/18970/219884893-ff39884b-0b4c-4c3c-80d9-97b7d0bf45d9.png)
![Screenshot 2023-02-18 at 18 12 53](https://user-images.githubusercontent.com/18970/219884896-5fc9d6ed-8505-47a6-b372-a615672f2fec.png)

(those are actual screenshots)

I'll note that the 3 apps have a visual VScode plugin, so getting this to work with two tabs side-by-side seems relatively straight-forward. I'm not sure about the feasibility of editing the sketch _directly within the document in visual mode_ (is this kind of interaction possible in VScode, mixed with regular text?). I'm also not sure of this for Rstudio's visual editor (it can display plots, equations, etc. but can an interactive widget be embedded?). Similar to VScode, the sketchpad could be displayed alongside in the html preview/plot pane (here's an [old experiment by timelyportfolio](https://github.com/timelyportfolio/excalidrawr))

**Comment by baptiste** [comment]
... and further down the line, one could imagine extending the concept to enable some interaction/two-way sync between sketches and code chunks, such as for example adding a hand-drawn annotation layer to plots or slides, etc. (and being able to edit/adjust this layer, programmatically or manually)

... and yet another step further down the line, one might imagine being able to interact directly with such sketches in the rendered document as well (Shiny, or Observable/d3, etc.), by targetting elements of the SVG. A classic use-case would be [something like this](http://worrydream.com/ExplorableExplanations/#explorableExample), where the reader can play with a schematic electrical circuit, and see the values in the text and in the plots update in real time.

**Comment by MurzNN** [comment]
Awesome idea, but sad that there is no progress here in 3 years :cry: 
About the unsynchronized content issue, we can store the source code of the embedded diagrams inline directly in the Quarto files, so each document version will have the correct embedded version, instead of referring to some online cloud storage, where the diagram can be changed at any time. Or, if the diagram size is pretty big, store it as attachment like images.

**Reply by mcanouil** [reply]
@MurzNN Try to keep your comments in one thread or eventually to edit them.

Note that Quarto extensions in exist, so you could try to make one.

**Comment by MurzNN** [comment]
Here https://github.com/JRJurman/excalidraw-to-svg I found a JS lib that is able to represent Excalidraw drawings as SVG. So, seems, using this we can render the Excalidraw drawings in the documents inline, like Mermaid.

**Reply by MurzNN** [reply]
Ouch, and there is an official library that can export to SVG and PNG, yay! https://www.npmjs.com/package/@excalidraw/utils

**Comment by MurzNN** [comment]
And here are the libraries for Draw.io (Diagrams.net) to convert it to SVG, PNG: 
https://www.npmjs.com/package/@whitebite/diagram-converter
https://www.npmjs.com/package/@mattiash/drawio-export
https://github.com/bearcove/home-drawio

**Comment by baptiste** [comment]
For what it's worth I'm relatively happy with the way VScode handles `.excalidraw.svg` files (svg file with embedded excalidraw scene); it lets me edit an sketch interactively and have it immediately updated in the document I'm writing. I mostly use this directly with typst these days, rather than via quarto, but the same idea would apply (though less instantaneous). Bonus of using typst is that it's also quite trivial to replace the font on the fly to match that used in the document.

<img width="1552" height="977" alt="Screenshot 2026-05-05 at 11 11 15" src="https://github.com/user-attachments/assets/18852995-cc36-425d-bbba-466f4cc07908" />
It's not quite embedded in the document itself as I was describing above, but a different pane of the same editor is just as good.

**Comment by MurzNN** [comment]
Found one more good solution to embed editable Draw.io (Diagrams.net) drawings and diagrams in Quarto documents in VS Code and derivatives using this extension: https://open-vsx.org/extension/hediet/vscode-drawio - it creates editable SVG and PNG files that can be embedded into Quarto Markdown files as regular images, but still be edited by the Draw.io editor directly in the VS Code interface, and stored as local files!
